tokio/src/runtime/thread_pool/queue.rs

```diff
@@ -0,0 +1,550 @@
+//! Run-queue structures to support a work-stealing scheduler
+
+use crate::loom::cell::{CausalCell, CausalCheck};
+use crate::loom::sync::atomic::{self, AtomicU32, AtomicUsize};
+use crate::loom::sync::{Arc, Mutex};
+use crate::task;
+
+use std::marker::PhantomData;
+use std::mem::MaybeUninit;
+use std::ptr::{self, NonNull};
+use std::sync::atomic::Ordering::{Acquire, Release};
+
+/// Producer handle. May only be used from a single thread.
+pub(super) struct Local<T: 'static> {
+    inner: Arc<Inner<T>>,
+
+    /// LIFO slot. Cannot be stolen.
+    next: Option<task::Notified<T>>,
+}
+
+/// Consumer handle. May be used from many threads.
+pub(super) struct Steal<T: 'static>(Arc<Inner<T>>);
+
+/// Growable, MPMC queue used to inject new tasks into the scheduler and as an
+/// overflow queue when the local, fixed-size, array queue overflows.
+pub(super) struct Inject<T: 'static> {
+    /// Pointers to the head and tail of the queue
+    pointers: Mutex<Pointers>,
+
+    /// Number of pending tasks in the queue. This helps prevent unnecessary
+    /// locking in the hot path.
+    ///
+    /// The LSB is a flag tracking whether or not the queue is open or not.
+    len: AtomicUsize,
+
+    _p: PhantomData<T>,
+}
+
+pub(super) struct Inner<T: 'static> {
+    /// Concurrently updated by many threads.
+    head: AtomicU32,
+
+    /// Only updated by producer thread but read by many threads.
+    tail: AtomicU32,
+
+    /// Elements
+    buffer: Box<[CausalCell<MaybeUninit<task::Notified<T>>>]>,
+}
+
+struct Pointers {
+    /// True if the queue is closed
+    is_closed: bool,
+
+    /// Linked-list head
+    head: Option<NonNull<task::Header>>,
+
+    /// Linked-list tail
+    tail: Option<NonNull<task::Header>>,
+}
+
+unsafe impl<T: task::ScheduleSendOnly> Send for Inner<T> {}
+unsafe impl<T: task::ScheduleSendOnly> Sync for Inner<T> {}
+unsafe impl<T: task::ScheduleSendOnly> Send for Inject<T> {}
+unsafe impl<T: task::ScheduleSendOnly> Sync for Inject<T> {}
+
+#[cfg(not(loom))]
+const LOCAL_QUEUE_CAPACITY: usize = 256;
+
+// Shrink the size of the local queue when using loom. This shouldn't impact
+// logic, but allows loom to test more edge cases in a reasonable a mount of
+// time.
+#[cfg(loom)]
+const LOCAL_QUEUE_CAPACITY: usize = 2;
+
+const MASK: usize = LOCAL_QUEUE_CAPACITY - 1;
+
+/// Create a new local run-queue
+pub(super) fn local<T: 'static>() -> (Steal<T>, Local<T>) {
+    debug_assert!(LOCAL_QUEUE_CAPACITY >= 2 && LOCAL_QUEUE_CAPACITY.is_power_of_two());
```

Comment:
Is there a way for us to make this a static assertion instead?

Reply:
do you have thoughts on how?

Reply:
https://github.com/nvzqz/static-assertions-rs/blob/v1.0.0/src/const_assert.rs

Reply:
Seems like hax 😆 you want me to lift that macro?